server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java

```diff
@@ -0,0 +1,230 @@
+                     : CompletableFuture.completedFuture(v));
+      }
+      return pollStage;
+   }
+
+   @Listener(clustered = true)
+   public static class PubSubListener {
+      private final Channel channel;
+      EmbeddedMultimapListCache<byte[], byte[]> multimapList;
+      public AdvancedCache<Object, Object> cache;
+      ScheduledFuture<?> scheduledTimer;
+      private boolean listenerAdded = false;
+      private CompletableFuture<Collection<byte[]>> listnerFuture = new CompletableFuture<>();
+      private CompletableFuture<Collection<byte[]>> pollFuture = new CompletableFuture<>();
+      private CompletableFuture<Collection<byte[]>> future;
+      Resp3Handler handler;
+
+      public void setListenerAdded(boolean listenerAdded) {
+         this.listenerAdded = listenerAdded;
+      }
+
+      public PubSubListener(Channel channel, Resp3Handler handler, AdvancedCache<Object, Object> cache,
+            EmbeddedMultimapListCache<byte[], byte[]> mml) {
+         this.channel = channel;
+         this.multimapList = mml;
+         this.cache = cache;
+         this.handler = handler;
+         // This future sync poll thread and listener. It waits for the poll stage completion
+         // and its value returned if available otherwise wait for events
+         future = pollFuture.thenCompose((v) -> v != null ? CompletableFuture.completedFuture(v) : listnerFuture)
+               .whenComplete((ignore_v, ignore_t) -> {
+                  try {
+                     this.deleteTimer();
+                     if (listenerAdded) {
+                        cache.removeListenerAsync(this);
+                     }
+                  } catch (Exception ex) {
+                     System.out.println(ex.toString());
+                  }
+               });
+      }
+
+      public CompletableFuture<Collection<byte[]>> getFuture() {
+         return future;
+      }
+
+      public void completePoll(Collection<byte[]> entry) {
+         pollFuture.complete(entry);
+      }
+
+      public void completeExceptionally(Throwable t) {
+         future.completeExceptionally(t);
+      }
+
+      public void startTimer(long timeout) {
+         deleteTimer();
+         scheduledTimer = (timeout > 0) ? handler.getScheduler().schedule(() -> {
+            if (listenerAdded) {
+               cache.removeListenerAsync(this);
+            }
+            future.complete(null);
+         }, timeout, TimeUnit.MILLISECONDS) : null;
+      }
+
+      public void deleteTimer() {
+         if (scheduledTimer != null)
+            scheduledTimer.cancel(true);
+         scheduledTimer = null;
+      }
+
+      @CacheEntryCreated
+      public CompletionStage<Void> onEvent(CacheEntryEvent<Object, Object> entryEvent) {
+         try {
+            if (entryEvent.getValue() instanceof ListBucket) {
+               byte[] key = unwrapKey(entryEvent.getKey());
+               multimapList.pollFirst(key, 1).handle((v, t) -> {
```

Comment:
This will still be a problem as we could get multiple events in the time before we remove this listener and thus it would cause multiple keys to be popped. I am thinking we need to fire BLPOP again honestly to remove values as necessary and then ensure with the original BLPOP that it doesn't fire an extra one if the second pollAllKeys found something already.

Also there should be some sort of AtomicBoolean to only fire this once.

Reply:
I see...
How about something like:
- onEvent side: if the second poll is ongoing save the first received key, else perform the pop and unsubscribe?
- on secondpoll side: on complete if a key has been found pop it, else if a key has been saved from onEvent pop it. unsubscribe

Reply:
@wburns I redesigned the code following the approach hinted above